app/notification.go

```diff
@@ -260,16 +296,119 @@ func (a *App) SendNotifications(post *model.Post, team *model.Team, channel *mod
+	var threadMentionedUserIds map[string]string
+
+	if post.Type == model.POST_ADD_TO_CHANNEL {
+		val := post.Props[model.POST_PROPS_ADDED_USER_ID]
+		if val != nil {
+			uid := val.(string)
+			m.MentionedUserIds[uid] = true
+		}
+	}
+
+	// get users that have comment thread mentions enabled
+	if len(post.RootId) > 0 && parentPostList != nil {
+		for _, threadPost := range parentPostList.Posts {
+			profile := profileMap[threadPost.UserId]
+			if profile != nil && (profile.NotifyProps[model.COMMENTS_NOTIFY_PROP] == THREAD_ANY || (profile.NotifyProps[model.COMMENTS_NOTIFY_PROP] == THREAD_ROOT && threadPost.Id == parentPostList.Order[0])) {
+				if threadPost.Id == parentPostList.Order[0] {
+					threadMentionedUserIds[threadPost.UserId] = THREAD_ROOT
+				} else {
+					threadMentionedUserIds[threadPost.UserId] = THREAD_ANY
+				}
+
+				if _, ok := mentionedUserIds[threadPost.UserId]; !ok {
+					mentionedUserIds[threadPost.UserId] = false
+				}
+			}
+		}
+	}
+
+	// prevent the user from mentioning themselves
+	if post.Props["from_webhook"] != "true" {
+		delete(mentionedUserIds, post.UserId)
+	}
+
+	if len(m.OtherPotentialMentions) > 0 && !post.IsSystemMessage() {
+		if result := <-a.Srv.Store.User().GetProfilesByUsernames(m.OtherPotentialMentions, &model.ViewUsersRestrictions{Teams: []string{team.Id}}); result.Err == nil {
+			channelMentions := model.UserSlice(result.Data.([]*model.User)).FilterByActive(true)
+
+			var outOfChannelMentions model.UserSlice
+			var outOfGroupsMentions model.UserSlice
+
+			if channel.GroupConstrained != nil && *channel.GroupConstrained {
+				nonMemberIDs, err := a.FilterNonGroupChannelMembers(channelMentions.IDs(), channel)
+				if err != nil {
+					return nil, nil, nil, err
+				}
+
+				outOfChannelMentions = channelMentions.FilterWithoutID(nonMemberIDs)
+				outOfGroupsMentions = channelMentions.FilterByID(nonMemberIDs)
+			} else {
+				outOfChannelMentions = channelMentions
+			}
+
+			if channel.Type != model.CHANNEL_GROUP {
+				a.Srv.Go(func() {
+					a.sendOutOfChannelMentions(sender, post, outOfChannelMentions, outOfGroupsMentions)
+				})
+			}
 		}
 	}
 
+	// find which users in the channel are set up to always receive mobile notifications
+	for _, profile := range profileMap {
+		if (profile.NotifyProps[model.PUSH_NOTIFY_PROP] == model.USER_NOTIFY_ALL ||
+			channelMemberNotifyPropsMap[profile.Id][model.PUSH_NOTIFY_PROP] == model.CHANNEL_NOTIFY_ALL) &&
+			(post.UserId != profile.Id || post.Props["from_webhook"] == "true") &&
+			!post.IsSystemMessage() {
+			allActivityPushUserIds = append(allActivityPushUserIds, profile.Id)
+		}
+	}
+	return allActivityPushUserIds, mentionedUserIds, channelMemberNotifyPropsMap, nil
+}
+
+// Send Push Notifications based on mentioned or active users
+func (a *App) sendPushNotifications(mentionedUsersList []string, profileMap map[string]*model.User, threadMentionedUserIds map[string]string, post *model.Post, notification *postNotification,
```

Comment:
This function also takes a lot of arguments. I'm not sure exactly how to handle this, but a lot of those arguments are just passed through to `sendOrRegisterNotification`, so I feel like there might be something we could do with that.

Reply:
We can wait until next iteration here too.